### ISSUE TITLE
Windows build

### ISSUE THREAD

**Issue by ocefpaf** [conda-forge/nco-feedstock#6]
We should build should build `nco` from source on Windows.

See http://nco.sourceforge.net/nco_qt_msvc.shtml

**Comment by ocefpaf** [comment]
Ping @czender. If you have any hints for us here it would really help us get the Windows builds going.

**Comment by czender** [comment]
the expert on windows builds of NCO is @pedro-vicente

**Comment by pedro-vicente** [comment]
@ocefpaf 

Building NCO from source is a time consuming task that for a first timer could take several frustrating days to finish. On the NCO main page you will see a list of library dependencies

http://nco.sourceforge.net/

on the link you posted above there are several environment variables listed

LIB_NETCDF
LIB_HDF5
LIB_HDF5_HL
LIB_ZLIB
LIB_SZIP
LIB_ANTLR
LIB_GSL
LIB_CURL
LIB_UDUNITS
LIB_EXPAT

Additionnaly, the following environment variable must also be defined for the location of the netCDF, UDUnits, GSL and Antlr header files:
HEADER_NETCDF
HEADER_UDUNITS
HEADER_GSL
HEADER_ANTLR

By their name it's obvious to which library they refer. So what you have to do is:

1) Obtain the source code for all the libraries above. (ANTLR must be version 2.7)
LIB_UDUNITS
LIB_EXPAT
these 2 it's better to send an email to Ward Fisher on the netcdf mailing list because I think the netcdf group ported a semi-official port of these for Windows (that by now might be official)

2) Build the libraries, and define the above environment variables

Boa sorte !

**Comment by ocefpaf** [comment]
Obrigado Pedro! I'll take give it a try. We have almost all of those in
conda-forge. I need to work on antlr though.
Em 13/06/2016 23:06, "Pedro Vicente" <notifications@github.com> escreveu:

**Comment by rsignell-usgs** [comment]
@ocefpaf, seems like recent progress allowing `antlr` window builds
https://github.com/antlr/antlr4/pull/1268

**Comment by ocefpaf** [comment]
> @ocefpaf, seems like recent progress allowing `antlr` window builds
https://github.com/antlr/antlr4/pull/1268

See my comment in https://github.com/conda-forge/pynco-feedstock/issues/2#issuecomment-251936898

I never got time to look at the Windows builds but it should be easier now that conda brings `msys2` in the default channel.

**Comment by czender** [comment]
that's good news but beware that NCO requires antlr version 2.7 not antlr version 4.x. not sure whether the antlr4 package will cut the mustard for NCO.

**Comment by kwilcox** [comment]
@ocefpaf  I have a windows laptop and an afternoon... if you want to give me a 10 minute brief I can hack away at it.

We can start with a non-ncap2 build for Windows and make incremental improvements as needed if its blocked upstream somewhere.

**Comment by ocefpaf** [comment]
@kwilcox I am a little bit swamped but you can start by setting your laptop up with conda and a MSVC.

https://github.com/conda/conda-recipes/wiki/Building-from-Source-on-Windows-32-bit-and-64-bit

Once that is done we need to check if we have all the dependencies already in conda and compiled on Windows.

- [X] LIB_NETCDF, HEADER_NETCDF
- [X] LIB_HDF5
- [ ] LIB_HDF5_HL  # not sure
- [X] LIB_ZLIB
- [ ] LIB_SZIP  # absent but I believe it is not mandatory (alternative to zlib)
- [ ] LIB_ANTLR, HEADER_ANTLR  # absent
- [ ] LIB_GSL, HEADER_GSL  # absent but I believe it is not mandatory (`ncap2`?)
- [X] LIB_CURL
- [X] LIB_UDUNITS, HEADER_UDUNITS
- [X] LIB_EXPAT

**Comment by czender** [comment]
you are correct that antlr and gsl are only required for ncap2, and that szip is not mandatory (NCO does not require it, though netCDF4 may require it depending on how it is built). so building NCO with --disable-ncap2 might be the easiest way to start.

**Comment by pedro-vicente** [comment]
LIB_HDF5_HL # not sure 

this is required , since netCDF uses the HDF5 High Level library 

LIB_SZIP # absent but I believe it is not mandatory (alternative to zlib) 

not mandatory, but if the HDF5 version was linked with it, then it is required , otherwise you will get linking errors

LIB_ANTLR, HEADER_ANTLR # absent 

optional, must be 2.7, we have a port for Windows in NCO

LIB_GSL, HEADER_GSL # absent but I believe it is not mandatory (ncap2?) 

optional, we have a port for Windows in NCO

-Pedro


  ----- Original Message -----

**Comment by rsignell-usgs** [comment]
@pedro-vicente, any chance you could take a stab at getting the settings right for a conda build on windows without ncap2?    

I might be able to help answer conda issues...

**Comment by ocefpaf** [comment]
Based on http://nco.sourceforge.net/nco_qt_msvc.shtml we need Qt's `qmake`. @ccordoba12 do you know if that is possible with `anaconda`'s `Qt`? Any experience with `qmake`?

@czender would it be complicated to switch to `cmake`?

@pedro-vicente if you want to try this please feel free to bug me with any `conda`/`conda-build` question.

**Comment by czender** [comment]
@ocefpaf A switch from autotools to cmake might be conceptually simple, yet I've never written cmake build scripts and my time is limited. If someone else provided a working cmake build system then I would be happy to include it in NCO with the intent of maintaining both cmake and autotools build systems so long as each has significant usage.

**Comment by ccordoba12** [comment]
`qmake` comes with all our Qt builds :-) I don't have much experience with it though.

**Comment by pedro-vicente** [comment]
@rsignell-usgs 
>>any chance you could take a stab at getting the settings right for a conda build on windows without >>ncap2? 

I never worked with conda, and my time is limited to learn yet another build system.
But if you post here or email me some instructions of where to start, I can give it a try, but no promises about time.

What are the problems you found on the conda build?

On my post above from June 13, there are the steps for building with the supplied NCO Qt build:
what you have to do is, with Visual Studio, build the NCO dependencies in some way.

I suggest to start with the minimum required, netCDF and HDF5. Both use Cmake,
so use Cmake to build these 

LIB_NETCDF (needs LIB_CURL)
LIB_HDF5 (needs LIB_ZLIB)
LIB_HDF5_HL

then define the above environment variables on Windows.

Here's the cmake command I used to build netCDF:

"C:\Program Files (x86)\CMake\bin\cmake" -G "Visual Studio 10" -C ../cmake/ConfigUser.cmake -D"NC_USE_STATIC_CRT=ON" -D"BUILD_SHARED_LIBS=OFF" -D"ENABLE_DLL=OFF" -D"HDF5_INCLUDE_DIR=K:\hdf5-1.8.13\src" -D"HDF5_LIB=K:\hdf5-1.8.13\build\bin\Debug\libhdf5_D.lib" -D"HDF5_HL_LIB=K:\hdf5-1.8.13\build\bin\Debug\libhdf5_hl_D.lib" -D"ZLIB_INCLUDE_DIR:PATH=K:\zlib\V1.2.5" -D"ZLIB_LIBRARY:FILEPATH=K:\zlib\V1.2.5\windows\static\zlib125d.lib"  -D"ENABLE_DAP=ON"  -D"CURL_LIBRARY=K:\curl-7.29.0\builds\libcurl-vc10-x86-debug-static-ipv6-sspi-spnego-winssl\lib\libcurl_a_debug.lib" -D"CURL_INCLUDE_DIR=K:\curl-7.29.0\builds\libcurl-vc10-x86-debug-static-ipv6-sspi-spnego-winssl\include"  ..

NetCDF depends on HDF5 and CURL
In this example
define these environment variables in Windows
LIB_HDF5 = K:\hdf5-1.8.13\build\bin\Debug\libhdf5_D.lib
LIB_CURL = K:\curl-7.29.0\builds\libcurl-vc10-x86-debug-static-ipv6-sspi-spnego-winssl\lib\libcurl_a_debug.lib

then define the  include path for netCDF as an environment variable
HEADER_NETCDF

if you go to the NCO distro, see the file /qt/nco.pri
It is part of the Qt build for NCO
The location of the dependencies are done this way

win32{
 netcdf3 {
 INCLUDEPATH += $(HEADER_NETCDF3)
 LIBS += $(LIB_NETCDF3)
 } else {
 INCLUDEPATH += $(HEADER_NETCDF)
 LIBS += $(LIB_NETCDF)
 LIBS += $(LIB_HDF5)
 LIBS += $(LIB_HDF5_HL)
 LIBS += $(LIB_ZLIB)
 LIBS += $(LIB_SZIP)
 DEFINES += ENABLE_NETCDF4
 DEFINES += HAVE_NETCDF4_H
}
 INCLUDEPATH += $(HEADER_GSL)
 INCLUDEPATH += $(HEADER_UDUNITS)
 LIBS += $(LIB_GSL)
 LIBS += $(LIB_UDUNITS)
 LIBS += $(LIB_EXPAT)
 LIBS += $(LIB_CURL)
}

then to build , we use at the command prompt

$ qmake

this can be done both for Unix and Windows. On Unix a Makefile is generated, on Windows,
Visual Studio projects are generated. 
NCO has actually 3 build systems for Unix: 1) automake, 2) a good old fashioned standalone Makefile at /bld/Makefile and 3) a Qt generated Makefile at /qt

The Qt build system is actually very simple, as you can see; it only needs hardcoded
include paths
INCLUDEPATH +
and hardcoded library files
LIBS +=

So, it does not try to "guess" anything , like Cmake does.

Cmake probably would be a better choice to build NCO, but it requires to build scripts that "guess" locations, and with many dependencies things can get complicated fast (see a recent thread of the PLplot ploting library for example, they use dozens of of .cmake scripts)

**Comment by ocefpaf** [comment]
> `qmake` comes with all our Qt builds :-) I don't have much experience with it though.

Thanks @ccordoba12! 

> I never worked with conda, and my time is limited to learn yet another build system.
But if you post here or email me some instructions of where to start, I can give it a try, but no promises about time.

I started a PR to show you what we can do so far. See https://github.com/conda-forge/nco-feedstock/pull/22

Conda is not that complicated as it is a layer on top of other build systems.

> On my post above from June 13, there are the steps for building with the supplied NCO Qt build:
what you have to do is, with Visual Studio, build the NCO dependencies in some way.

We have "cmake" builds for all but ANTLR and GSL.

https://github.com/conda-forge/hdf5-feedstock/blob/master/recipe/bld.bat
https://github.com/conda-forge/libnetcdf-feedstock/blob/master/recipe/bld.bat
https://github.com/conda-forge/zlib-feedstock/blob/master/recipe/bld.bat
https://github.com/conda-forge/curl-feedstock/blob/master/recipe/bld.bat
https://github.com/conda-forge/udunits2-feedstock/blob/master/recipe/bld.bat
https://github.com/conda-forge/expat-feedstock/blob/master/recipe/bld.bat

not sure how well those will play with `nco` but we can make adjustments to those recipes if necessary.

**Comment by pedro-vicente** [comment]
I did a [EDIT] of my previous reply above

**Comment by pedro-vicente** [comment]
@ocefpaf 

>>We have "cmake" builds for all but ANTLR and GSL.

Do the above bld.bat files get the libraries from some "official" location (like github) or do you host them somewhere?

I believe there is no "official" cmake script for Windows for  ANTLR 2.7 and GSL that you could do the same.

So, in this case you have to build  ANTLR 2.7 and GSL manually with compiler settings the same as the above bld.bat, and write a similar bld.bat for both in the conda build.

I have Visual Studio 2013 solutions to build  ANTLR 2.7 and GSL, but I don't have any way to make these available, like a public machine for download.

**Comment by ocefpaf** [comment]
> Do the above bld.bat files get the libraries from some "official" location (like github) or do you host them somewhere?

All official sources. In some rare cases we patch them to enable the build.

> I believe there is no "official" cmake script for Windows for ANTLR 2.7 and GSL that you could do the same.

That is in the background for now. We are focused on getting a build without `GSL` and `ANTLR`.

> I have Visual Studio 2013 solutions to build ANTLR 2.7 and GSL, but I don't have any way to make these available, like a public machine for download.

You can send it via e-mail or upload right here on the GitHub issues. (Drag-and-drop should work for most file formats.) BTW people discussing the `GSL` Windows build at https://github.com/conda-forge/gsl-feedstock/issues/7 will be quite happy with that :smile:

**Comment by pedro-vicente** [comment]
@ocefpaf 
@czender 
I volunteer to write a CMake build.

**Comment by czender** [comment]
Much appreciated, @pedro-vicente

**Comment by ocefpaf** [comment]
> I volunteer to write a CMake build.

that is awesome @pedro-vicente! Please feel free to bug me for any help you need.

**Comment by pedro-vicente** [comment]
The Cmake build is only for after NCO 4.6.2

**Comment by ocefpaf** [comment]
> The Cmake build is only for after NCO 4.6.2

Was that released? If not we can open the PR with a pre-release just to get the ball rolling.

**Comment by pedro-vicente** [comment]
NCO release dates are here
http://nco.sourceforge.net/

**Comment by czender** [comment]
Latest version is 4.6.2-beta03. Final 4.6.2 probably this coming week.

**Comment by pedro-vicente** [comment]
I pushed to github a prototype for a CMake build. The file /doc/index.shtml contains build instructions, which are

CMake build. To build with CMake, do:
mkdir build
cd build
cmake ..
make

The CMake script tries to find header and dependency libraries on standard system locations. If these are not found, they can be set with (the locations in full path are examples)

cmake .. -DNETCDF_INCLUDE:PATH=/my/netcdf/include/path -DNETCDF_LIBRARY=/my/netcdf/library/file -DHDF5_LIBRARY=/my/hdf5/library/file -DHDF5_HL_LIBRARY=/my/hdf5 high level/library/file -DSZIP_LIBRARY=/my/szip/library/file -DZLIB_LIBRARY=/my/zlib/library/file -DCURL_LIBRARY=/my/curl/library/file

At the moment only a linux build of ncks is attempted, and in CentOS with custom locations I get a couple of link errors

Right now only the minimal dependencies are detected: netcdf header file, netcdf, HDF5, HDF5_HL, zlib, szip, curl libraries

**Comment by czender** [comment]
Where does this issue stand? @pedro-vicente have you produced a working feedstock that builds a conda-installable NCO (including basic ncap2) on Windows?

**Comment by pedro-vicente** [comment]
@czender The cmake build status is the same as the above post.
Binaries are built for these NCO operators with just the minimum dependencies (netCDF, HDF5, szip, zlib)
ncks 
ncbo
ncecat
ncflint 
ncpdq 
ncra 
ncrename 

Regarding the conda build, can anyone please remind me what is needed? 
Is it a NCO cmake build, including library dependencies?
Like I said before, I can spend some time on the NCO cmake build but I don't have time to learn conda.

**Comment by czender** [comment]
That's a great start. I see the CMakelists.txt file in the top level directory. I think @ocefpaf or someone else familiar with the conda Windows environment ( @rsignell-usgs ? ) needs to give that a try and work it from the feedstock end, and tell us what mods they need in the upstream. @pedro-vicente please install conda on your windows machine so you can iterate/verify that the install works.

**Comment by ocefpaf** [comment]
> That's a great start. I see the CMakelists.txt file in the top level directory.

Sorry. I am a bit lost. You mean a custom  `CMakelists.txt` file somewhere or are you referring to the one shipped with the source code? (Sorry I never looked into that part of the code so maybe this is something obvious.)

> I think @ocefpaf or someone else familiar with the conda Windows environment ( @rsignell-usgs ? ) needs to give that a try and work it from the feedstock end, and tell us what mods they need in the upstream.

I am in the processing of moving to another city and I won't be able to try that for at least 2 weeks (sorry).

> @pedro-vicente please install conda on your windows machine so you can iterate/verify that the install works.

@pedro-vicente conda is not that complicated. I will get a simple PR that tries to build this so we can iterate together.

**Comment by pedro-vicente** [comment]
@ocefpaf 
I you do git pull of NCO, there is a CMakelists.txt file in the top level directory.
Do
mkdir build
cd build
cmake ..

If the dependencies header files/libraries are not found they can be with these options

NETCDF_INCLUDE
NETCDF_LIBRARY
HDF5_LIBRARY
HDF5_LIBRARY
HDF5_HL_LIBRARY
ZLIB_LIBRARY
SZIP_LIBRARY
CURL_LIBRARY

**Comment by ocefpaf** [comment]
`conda-forge` does not package `SZIP` because of its license. How can I bypass that in the `cmake` call?

```
CMake Error at CMakeLists.txt:62 (message):
  szip library not found
```

**Comment by czender** [comment]
FYI it is possible (and quite common) to build netCDF without SZIP, so that NCO will never need to link to SZIP. It should be possible to either eliminate all references to SZIP from CMakelists.txt, or to ignore SZIP when the package is not found.

**Comment by ocefpaf** [comment]
> it is possible (and quite common) to build netCDF without SZIP, so that NCO will never need to link to SZIP.

Our `netCDF` is not built with `szip` and `cmake` is still trying to find it.

> It should be possible to either eliminate all references to SZIP from CMakelists.txt, or to ignore SZIP when the package is not found. 

I see that in sometimes `SZIP=0` does the trick, would that work with `nco`? I want to avoid patching `CMakelists.txt` for `off`/`on` flags like that.

**Comment by czender** [comment]
Yes. The conda-forge Windows build is the raison d'etre for the CMakelists.txt. So please just submit any desired changes you need as pull-requests and @pedro-vicente or I will quickly merge them. After conda-forge Windows works, we can go back and improve CMakelists.txt for non-conda situations.

**Comment by ocefpaf** [comment]
> The conda-forge Windows build is the raison d'etre for the CMakelists.txt. So please just submit any desired changes you need as pull-requests and @pedro-vicente or I will quickly merge them. After conda-forge Windows works, we can go back and improve CMakelists.txt for non-conda situations.

Got it. I though it was already "generic." I will play with it here then and submit any patches if I later.

**Comment by monocongo** [comment]
I have run cmake on Windows, below is the output. Are there next steps?


"C:\User Programs\CLion\CLion 2016.3.2\bin\cmake\bin\cmake.exe"
-DCMAKE_BUILD_TYPE=Debug -G "CodeBlocks - MinGW Makefiles" C:\home\git\nco
-- Found netcdf.h header file at: C:/home/Anaconda3/Library/include
-- Found netcdf library at: C:/home/Anaconda3/Library/lib/netcdf.lib
-- Found hdf5 library at: C:/Anaconda3/Library/lib/hdf5.lib
-- Found hdf5 high level library at: C:/Anaconda3/Library/lib/hdf5_hl.lib
-- Found zlib library at: C:/Anaconda3/Library/lib/z.lib
-- Found curl library at: C:/Anaconda3/Library/lib/libcurl.lib
-- Configuring done
-- Generating done
-- Build files have been written to: C:/home/git/nco/cmake-build-debug

**Comment by pedro-vicente** [comment]
@ocefpaf 

>> szip library not found

To avoid this error , you can, for now

1) get SZIP , build, install, from here

https://github.com/erdc-cm/szip

2) comment in your local copy of CMakelists.txt the lines regarding SZIP detection

Because NCO has many possible combinations of library dependencies,
I wanted to keep the dependencies as much simple as possible, so for now I assumed all base libraries are required.
Ideally cmake would detect if SZIP is needed by inspecting HDF5 (it's HDF5 that uses SZIP, not netCDF).

for now I am going to change the SZIP detection to a warning in the script. 
Of course if SZIP is required, then the build will fail. 
So my preferred solution would be just to install SZIP, even if it's not needed.

I would like to be the only maintainer of the cmake script, so if you have any change to propose, post here, or email me directly

**Comment by pedro-vicente** [comment]
@monocongo 
ok, the generation of the build files succeeded.
Now, only thing to  do is build :-)

>>> -G "CodeBlocks - MinGW Makefiles"

I never used CodeBlocks, but this should have created a CodeBlocks project or a file called "Makefile".
If you have a "Makefile" just do 

make

most people use Visual Studio

cmake -G "Visual Studio 14" ..

this creates a Visual Studio solution file and projects

**Comment by ocefpaf** [comment]
> To avoid this error , you can, for now
1. get SZIP , build, install, from here

Like I mentioned in https://github.com/conda-forge/nco-feedstock/issues/6#issuecomment-278359639 `SZIP` is a no-no in `conda-forge` because it does not allowed for binary re-distribution. However, I did check the page to see if that changed in the past years and I found this statement in http://www.compressconsult.com/szip/:

>  The sourcecode for szip is here. There is a general public license for distribution and usage of szip included in the package, so no need to worry.

But I downloaded the source and could not find a license file stating that :unamused: so that is still a  no-no.

> 2. comment in your local copy of CMakelists.txt the lines regarding SZIP detection

Did that in https://github.com/ocefpaf/nco-feedstock/commit/e76fa6e94896cde3a754b1b697b8c43d3767a93a with no avail.

See the full log in: https://ci.appveyor.com/project/ocefpaf/nco-feedstock/build/1.0.27/job/0x1ddiggwwky1j6d

> I would like to be the only maintainer of the cmake script, so if you have any change to propose, post here, or email me directly

I prefer open development on GitHub with PRs and reviews instead of message posts and e-mails. Once we achieve a successful build here I can send a PR to `nco`'s repository with a configuration for continuous integration with AppVeyor to test the Windows builds at every new PR.

**Comment by pedro-vicente** [comment]
The CMake build was officially released.
Notes here
http://nco.sourceforge.net/#cmake

**Comment by czender** [comment]
@ocefpaf @rsignell-usgs Thanks to @pedro-vicente, NCO 4.6.9, just released contains a functional CMake build that works on Windows, and szip can be omitted to avoid license issues. Build can be as simple as 
```
cd nco/cmake
cmake .. -DFIND_SZIP=OFF -DCMAKE_INSTALL_PREFIX=/usr/local
```
or with fully specified paths to some libs, e.g.,

```cmake .. -DFIND_SZIP=OFF -DCMAKE_INSTALL_PREFIX=${HOME} -DNETCDF_INCLUDE:PATH=${NETCDF_INC} -DNETCDF_LIBRARY:FILE=${NETCDF_LIB}/libnetcdf.a -DHDF5_LIBRARY:FILE=${NETCDF_LIB}/libhdf5.a -DHDF5_HL_LIBRARY:FILE=${NETCDF_LIB}/libhdf5.a```

or see example [build.bat](https://github.com/nco/nco/tree/master/cmake/build.bat) for a Window-ish environment. Hope someone volunteers to create a new PR based on CMake for the conda feedstock for Windows.

**Comment by pedro-vicente** [comment]
@czender little correction :-) The SZIP and ZLIB detection are now automatic, so -DFIND_SZIP=OFF was removed.

the script CMakeLists.txt includes comments on how this is done. 
NOTE, in particular for Windows

```
#HDF5 can be optionally linked with the SZIP library (Science Data Lossless Compression Program) and ZLIB
#Symbol to detect in ZLIB can be only H5Z_DEFLATE, a structure only defined in H5Zdeflate.c if the filter is enabled
#For SZIP the structure can be only H5Z_SZIP, defined in H5Zszip.c if the filter is enabled
#check_library_exists() tries to link a temporary program with these symbols
#On MSVC for this detection to work, the HDF5 library MUST HAVE as additional dependencies the ZLIB and SZIP libraries,
#which is not a requirement for the library to build successfully


```

**Comment by rsignell-usgs** [comment]
There is also this drop in replacement for SZIP

https://www.dkrz.de/redmine/projects/aec/wiki/SZIP

**Comment by pedro-vicente** [comment]
Just to add to the SZIP/ZLIB issue. This is an example output one gets after doing

`cmake ..`

 regarding a needed ZLIB


```
-- Detecting if HDF5 library /usr/local/lib/libhdf5.so needs the ZLIB library...
-- Looking for H5Z_DEFLATE in /usr/local/lib/libhdf5.so
-- Looking for H5Z_DEFLATE in /usr/local/lib/libhdf5.so - found
-- ZLIB library is needed...
```

and here and example for SZIP detection

```
-- Detecting if HDF5 library /usr/local/lib/libhdf5.so needs the SZIP library...
-- Looking for H5Z_SZIP in /usr/local/lib/libhdf5.so
-- Looking for H5Z_SZIP in /usr/local/lib/libhdf5.so - not found
-- SZIP library is not needed...
```

For this example, Cmake tries to
1) find a ZLIB library, if not found exit with fatal error; if found, add it (e.g libz.a) to the list of libraries to link
2) for the SZIP case, nothing is done 

The inclusion of SZIP/ZLIB is pretty much an individual choice for who builds the *HDF5* library (e.g linux packages distributions)
Downstream users, starting with netCDF and ending with NCO have no control over this.

So, the option -DFIND_SZIP=<ON/OFF> made no sense, it was just a temporary manual fix because there was no automatic detection.

**Comment by ocefpaf** [comment]
See https://github.com/conda-forge/nco-feedstock/pull/43#issuecomment-331497257 for the latest attempt.
Seems to get quite far but it is failing with:

```
[ 96%] Linking C executable ncra.exe
failed to create symbolic link 'ncea': No error
```

**Comment by pedro-vicente** [comment]
@ocefpaf 

I am not sure why that error happens (maybe because it says "No error" :-))

but I did some changes to that part
do a new pull from git hub an try again

It seems the error is on the build part

on the configure part, this should print

cmake ..

`-- Installing links ncea;nces;ncrcat;ncdiff;ncunpack to: /home/pvicente/nco_install/bin`

and on the build

make install

```
[ 71%] Linking C executable ncra
Making symbolic link /home/pvicente/nco_install/bin/ncra -> ncea
```

**Comment by pedro-vicente** [comment]
ok I see

```
[ 96%] Linking C executable ncra.exe
failed to create symbolic link 'ncea': No error
```

this was in Windows. This fails in Windows because Cmake's create_symlink is not for Windows.
Some days ago I added a 

`if (UNIX)`

test so that that error does not happen in Windows. It seems you are still using an old version before that change

**Comment by ocefpaf** [comment]
I am using the latest release. I guess the changes you are referring to did
not make into the release then.

**Comment by czender** [comment]
@pedro-vicente I am preparing to release 4.7.0. Any testing you can do to ensure that CMake build (still) works with Windows would be appreciated. Hopefully everything is set so that the conda feedstock will "just work" with Windows as of 4.7.0.

**Comment by pedro-vicente** [comment]
@czender CMake/NCO builds for 4.7.0

**Comment by xylar** [comment]
I have a PR #47 in for updating to 4.7.0.  Are there additional changes needed to add windows support to the recipe?  Maybe that can be a follow-up PR, if so.

**Comment by ocefpaf** [comment]
I have a PR with the RC that has the changes. I'll submit it tomorrow. It is "kind of" passing, the build completes but a `ncks -m $url` test fails :unamused:

**Comment by czender** [comment]
FWIW that test is to see whether the DAP protocol works. The test executes
`ncks -m http://tds.marine.rutgers.edu/thredds/dodsC/roms/espresso/2013_da/his/ESPRESSO_Real-Time_v2_History_Best`
That dataset currently exists, and the test passes on all my Linux/Mac boxen, yet the location is not robust since no one knows who controls that DAP server and when it may disappear. It would be better to test a dataset in a more reliable location so that a change on the Rutgers server does not break the conda builds.

**Comment by xylar** [comment]
Indeed, it seems like someone moved or deleted the dataset?

**Comment by xylar** [comment]
Hmm, works through ncks but if I try to access the URL with a browser it gives me an error:
```
Error {
    code = 400;
    message = "Unrecognized request";
};
```
Maybe this is unrelated to the test failing.

**Comment by czender** [comment]
The dataset can only be seen through the DAP protocol that ncks tries to invoke, and it appears to be broken with the Windows build only. This is not a deal-breaker as the rest of NCO is unaffected, but it would be nice to understand why DAP fails this test with Windows only. DAP could easily be misconfigured, e.g., if the build machine has a .daprc or a .dodsrc file it could be misdirecting the DAP request...

**Comment by pedro-vicente** [comment]
this

ncks -m http://tds.marine.rutgers.edu/thredds/dodsC/roms/espresso/2013_da/his/ESPRESSO_Real-Time_v2_History_Best

is working fine on Windows

netcdf ESPRESSO_Real-Time_v2_History_Best {
  dimensions:
    boundary = 4 ;
    eta_psi = 81 ;
    eta_rho = 82 ;
....

**Comment by ocefpaf** [comment]
> is working fine on Windows

I meant our build on AppVeyor.

> Hmm, works through ncks but if I try to access the URL with a browser it gives me an error:

@xylar you need to append `.html` if you want to open in a browser, but the `400` return means that there is something there, only the browser was not able to use it.

Here are the [AppVeyor build logs](https://ci.appveyor.com/project/ocefpaf/nco-feedstock/history) for those interested, and [this](https://github.com/ocefpaf/nco-feedstock/tree/try_win) is the branch I am working on. I'll update it to use the latest release soon, right now it is building from the rc release.

**Comment by ocefpaf** [comment]
@czender I believe the [`bld.bat`](https://github.com/ocefpaf/nco-feedstock/blob/try_win/recipe/bld.bat) for that build can help identify the issues. However, some packages get found anyway, like `antlr`.

**Comment by czender** [comment]
@ocefpaf what makes you think ANTLR is "found anyway"? It's not requested in bld.bat, and the AppVeyor build log quoted above shows that it is not found. And because it is not found, building ncap2 is not attempted. At least that's how I read the log. Here is the CMake example file, `nco/cmake/build.bat`, distributed with NCO:
```
cmake .. ^
-DNETCDF_INCLUDE:PATH=I:/grafic_project/netcdf/include ^
-DNETCDF_LIBRARY:FILE=I:/grafic_project/Debug/netcdf.lib ^
-DHDF5_LIBRARY:FILE=I:/grafic_project/Debug/hdf5.1.8.5.patch1.lib ^
-DHDF5_HL_LIBRARY:FILE=I:/grafic_project/Debug/hdf5_hl.1.8.5.patch1.lib ^
-DZLIB_LIBRARY:FILE=I:/grafic_project/Debug/zlib.lib ^
-DSZIP_LIBRARY:FILE=I:/grafic_project/Debug/szlib.lib ^
-DCURL_LIBRARY:FILE=I:/grafic_project/Debug/libcurld.lib ^
-DANTLR_INCLUDE:PATH=I:/antlr-2.7.7/lib/cpp ^
-DANTLR_LIBRARY:FILE=I:/antlr-2.7.7/lib/cpp/debug/antlr.lib ^
-DUDUNITS2_INCLUDE:PATH=I:/udunits2/lib ^
-DUDUNITS2_LIBRARY:FILE=I:/udunits2/build/lib/Debug/udunits2.lib ^
-DEXPAT_LIBRARY:FILE=I:/expat/expat/win32/bin/Debug/libexpatMT.lib ^
-DGSL_INCLUDE:PATH=I:/gsl-1.8/src ^
-DGSL_LIBRARY:FILE=I:/gsl-1.8/src/gsl/1.8/gsl-1.8/VC8/libgsl/Debug-StaticLib/libgsl_d.lib ^
-DGSL_CBLAS_LIBRARY:FILE=I:/gsl-1.8/src/gsl/1.8/gsl-1.8/VC8/libgslcblas/Debug-StaticLib/libgslcblas_d.lib
```
I suggest adding the equivalent paths for ANTLR to bld.bat that will work on AppVeyor.

**Comment by ocefpaf** [comment]
> @ocefpaf what makes you think ANTLR is "found anyway"?

A typo. I meant not found :smile: (because we don't have that for windows.)

**Comment by ocefpaf** [comment]
> There should be ANTLR version 2.7.7 source code available somewhere

The source is available and we do [build it](https://github.com/conda-forge/antlr-feedstock/) for OS X and Linux. We would love to add Windows builds we do lack the know-how there.

> I'm not sure about the best way to make this public. 

Using a pre-compiled binary blob may be an option, but I would like to avoid that if possible.
If we cannot build it the alternative would be to re-package your binary in the conda-forge channel.

**Comment by pedro-vicente** [comment]
@ocefpaf if you have the source, then all is needed are the Visual Studio projects. I'll attach those somewhere here later. I'll give it a try your bld.bat
Using a pre-compiled binary blob is a bad idea

**Comment by pedro-vicente** [comment]
here it is the Visual Studio project
this should located in
/lib/cpp
on your ANTLR source
I remember there was a compiling error using the Linux only official source, if you have that, post here, but it was a trivial thing
[antlr.zip](https://github.com/conda-forge/nco-feedstock/files/1458984/antlr.zip)

**Comment by ocefpaf** [comment]
Thanks. I'll give it a try over the weekend.

**Comment by pedro-vicente** [comment]
@ocefpaf I I tried on the master branch of nco-feedstock

```
K:\nco-feedstock>conda install nco
Fetching package metadata ...............

PackageNotFoundError: Packages missing in current channels:

  - nco

We have searched for the packages in the following channels:

  - https://conda.anaconda.org/conda-forge/win-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.continuum.io/pkgs/main/win-64
  - https://repo.continuum.io/pkgs/main/noarch
  - https://repo.continuum.io/pkgs/free/win-64
  - https://repo.continuum.io/pkgs/free/noarch
  - https://repo.continuum.io/pkgs/r/win-64
  - https://repo.continuum.io/pkgs/r/noarch
  - https://repo.continuum.io/pkgs/pro/win-64
  - https://repo.continuum.io/pkgs/pro/noarch
  - https://repo.continuum.io/pkgs/msys2/win-64
  - https://repo.continuum.io/pkgs/msys2/noarch
```

**Comment by ocefpaf** [comment]
> @ocefpaf I tried on the master branch of nco-feedstock

Not sure what you mean by that. The PR is in #48, not master.

Also, there won't be a package in the channel until we merge that PR.
In the PR we have all the information for the test build and the logs.

**Comment by pedro-vicente** [comment]
@ocefpaf I tried the try_win branch, I get the same error

**Comment by ocefpaf** [comment]
> @ocefpaf I tried the try_win branch, I get the same error

Same error as the AppVeyor logs? However, if you are referring the the package not found above I guess we still have a confusion.

The built package will be available in the channel for installation via  `conda install nco` only **after** we merge the PR that submits it, like #48.

That means, we need to make #48 work first. One can test the PR locally by forking/cloning the repo, getting that particular PR, and issue the build commands locally.

**Comment by pedro-vicente** [comment]
@ocefpaf

what I did

git clone https://github.com/conda-forge/nco-feedstock
git checkout try_win

I installed Anaconda3, opened a  Anaconda3 command prompt
then did what it says here

https://github.com/conda-forge/nco-feedstock

conda config --add channels conda-forge
conda install nco

error is the one above

PackageNotFoundError: Packages missing in current channels:

  - nco

**Comment by pedro-vicente** [comment]
@ocefpaf 

> The built package, like I said in #6 (comment), will be available in the channel for installation, via conda install nco, only after we merge the PR that submits it, like #48.
> 
> That means, we need to make #48 work first. One can test the PR locally by forking/cloning the repo, getting that particular PR, and issue the build commands locally.

It would be interesting to find out why that ncks -M test fails, but that shouldn't prevent you from release the nco-feedstock as it is, after adding ANTLR. 

If all nco-feedstock builds that should be fine. We NCO developers can then try to find the cause of that error. So, I would just remove that test, add ANTLR and release it

**Comment by ocefpaf** [comment]
> what I did

``` shell 
git clone https://github.com/conda-forge/nco-feedstock
git checkout try_win
```

OK you got the code to build the package, but the [instructions](https://github.com/conda-forge/nco-feedstock#installing-nco) only refer to built packages. `nco` in that PR is not built, it is "building" :wink: 
Only after the merge it will be available.

> It would be interesting to find out why that ncks -M test fails, but that shouldn't prevent you from release the nco-feedstock as it is, after adding ANTLR.

Indeed. If you want to try it locally you are close. Now that you have the code for #48 you can issue:

```
conda build recipe
```

from the code main directory. Building on Windows is kind of annoying and that probably won't succeed in the first try. You may need to configure your channels to get the dependencies from conda-forge (I guess that was already done above when you tried to install `nco`) . There may be also some trouble configuring MSVC. If you get a "`vcvarsall.bat`  not found" message you may need to call the problem `.bat` manually b/c conda was unable to find it.

> So, I would just remove that test, add ANTLR and release it

The only test passing is `ncks --help`, so I don't think that is advisable. I want to do at least a local file test before merging that PR. Also, I'll merge that only after #46 and #47.

> If all nco-feedstock builds that should be fine. We NCO developers can then try to find the cause of that error.

Sure. If you need more help setting a conda environment for building and testing it locally I am glad to help. Although Windows is not my forte :grimacing: (I rely mostly on AppVeyor for that.)

**Comment by pedro-vicente** [comment]
@ocefpaf 

`The only test passing is ncks --help, so I don't think that is advisable. I want to do at least a local file test before merging that PR. `

agreed. @czender probably is more qualified to choose a test. 
You happened to choose a problematic/difficult to debug test, because it relies on DAP, curl, the internet. Basically all NCO tests do some change to a local file, and the read the result to check if it is what expected.

**Comment by pedro-vicente** [comment]
@ocefpaf 

> Sure. If you need more help setting a conda environment for building and testing it locally I am glad to help. Although Windows is not my forte

ok

I did 

```
conda build recipe
Skipped: K:\nco-feedstock\recipe defines build/skip for this configuration.
```

and nothing seemed to happen 

How can I use AppVeyor?

**Comment by ocefpaf** [comment]
If you got `K:\nco-feedstock\recipe defines build/skip for this configuration.` that means you are not using the `try_win` branch. The main branch skips Windows.

AppVeyor will build automatically when pushing a PR, so nothing different from what we have now.

> You happened to choose a problematic/difficult to debug test, because it relies on DAP, curl, the internet. 

Yep. Package testing is like integration tests, not really meant for debugging, only to check if the expected features are working.

**Comment by pedro-vicente** [comment]
@ocefpaf I am totally sure I am using try_win :smiley:

can I change that setting in one of the files?

also, do I need to change this ?

> CONDA_INSTALL_LOCN: C:\\Miniconda36-x64

**Comment by ocefpaf** [comment]
> @ocefpaf I am totally sure I am using try_win :smiley:

Are you on a [Windows 32 machine or Python <=35](See https://github.com/conda-forge/nco-feedstock/pull/48/files#diff-e178b687b10a71a3348107ae3154e44cR13)?  If the former I don't think the recipe will work. If the latter than you can probably just remove that line b/c it has no baring outside of `conda-forge` (Python is attached to the msvc and both Python 3.5 and 3.6 use vc14, so we need to build only once and we choose to build with conda on Python 3.6).

**Comment by czender** [comment]
Regarding question above about a simple local file test, this may suffice and does not require ncap2:
```
fin=data/in.nc
ncks -H --trd -v one $fin
# Success: output to stdout is "one = 1" and zero return code
# Failure: non-zero return code and lots of error messages
```

**Comment by ocefpaf** [comment]
Thanks @czender. I'll add it tomorrow. #47 is almost ready and it s time to :sleeping: here.

PS: most of the delay is Travis-CI backlog to build the OS X version :unamused:

**Comment by pedro-vicente** [comment]
@ocefpaf 

`Are you on a Windows 32 machine or Python <=35?`

I have Windows 10 64 bit and I installed Anaconda 32 bit with Python 3.6. Should I try with Anaconda 64 bit.?

**Comment by pedro-vicente** [comment]
@ocefpaf 
that was it, with  Anaconda 64 bit it builds

**Comment by pedro-vicente** [comment]
I found the cause of the error, that turns out to be a NCO bug

@ocefpaf
if you want to reproduce this or any other bug do

1) change the cmake target to be 

`cmake -G "Visual Studio 14 Win64"`

so that it generates a Visual Studio solution for debugging

2) open the solution in Visual Studio and edit a project (say ncks) Properties->Debugging->Environment

add the PATH where the .DLLs for dependent libraries are located , e.g

`PATH=%PATH%;C:\Users\pedro\Anaconda3\conda-bld\nco_1510360320825\_h_env\Library\bin;C:\Users\pedro\Anaconda3\conda-bld\nco_1510360320825\_h_env\Scripts`

by the way is it possible to have static linking (.lib instead of .dll) of all the dependency libraries?

**Comment by pedro-vicente** [comment]
@ocefpaf
Is it possible to use the current NCO git code into the nco-feedstock? the current code should fix the Win64 issue that made your test fail.

**Comment by ocefpaf** [comment]
> @ocefpaf
Is it possible to use the current NCO git code into the nco-feedstock? the current code should fix the Win64 issue that made your test fail.

It is possible. See https://github.com/conda-forge/nco-feedstock/pull/48/commits/cdbcba99cb0fedf9e6760d2f8f5609ffd7ae8919

**Comment by pedro-vicente** [comment]
there is a build error because the file AST.hpp is missing from

 -D ANTLR_INCLUDE:PATH=C:\Users\pedro\Anaconda3\conda-bld\nco_1513179527208\_h_env\Library\include\antlr  


[ 79%] Building CXX object CMakeFiles/ncap2.dir/src/nco++/Invoke.cc.obj
Invoke.cc
C:\Users\pedro\Anaconda3\conda-bld\nco_1513179527208\work\src\nco++\Invoke.cc(11): fatal error C1083: Cannot open include file: 'antlr/AST.hpp': No such file or directory

**Comment by ocefpaf** [comment]
Done in #48
